src/lib.rs

```diff
@@ -240,71 +269,81 @@ impl Socket {
     /// server responded in a given timespan. This message takes an event, which
     /// could either be one of the common events like "message" or "error" or a
     /// custom event like "foo", as well as a data parameter. But be careful,
-    /// the string needs to be valid JSON. It's even recommended to use a
-    /// library like serde_json to serialize the data properly. It also requires
-    /// a timeout `Duration` in which the client needs to answer. This method
-    /// returns an `Arc<RwLock<Ack>>`. The `Ack` type holds information about
-    /// the `ack` system call, such whether the `ack` got acked fast enough and
-    /// potential data. It is safe to unwrap the data after the `ack` got acked
-    /// from the server. This uses an `RwLock` to reach shared mutability, which
-    /// is needed as the server sets the data on the ack later.
+    /// in case you send a [`Payload::String`], the string needs to be valid JSON.
+    /// It's even recommended to use a library like serde_json to serialize the data properly.
+    /// It also requires a timeout `Duration` in which the client needs to answer.
+    /// If the ack is accked in the correct timespan, the specified callback is
+    /// called. The callback consumes a [`Payload`] which represents the data send
+    /// by the server.
     ///
     /// # Example
     /// ```
-    /// use rust_socketio::SocketBuilder;
+    /// use rust_socketio::{SocketBuilder, Payload};
     /// use serde_json::json;
     /// use std::time::Duration;
     /// use std::thread::sleep;
     ///
     /// let mut socket = SocketBuilder::new("http://localhost:4200")
-    ///     .on("foo", |str| println!("Received: {}", str))
+    ///     .on("foo", |payload: Payload| println!("Received: {:#?}", payload))
     ///     .connect()
     ///     .expect("connection failed");
     ///
     ///
-    /// let payload = json!({"token": 123});
-    /// let ack_callback = |message| { println!("{}", message) };
     ///
-    /// socket.emit_with_ack("foo", &payload.to_string(),
-    /// Duration::from_secs(2), ack_callback).unwrap();
+    /// let ack_callback = |message: Payload| {
+    ///     match message {
+    ///         Payload::String(str) => println!("{}", str),
+    ///         Payload::Binary(bytes) => println!("Received bytes: {:#?}", bytes),
+    ///    }    
+    /// };
+    ///
+    /// let payload = json!({"token": 123});
+    /// socket.emit_with_ack("foo", payload, Duration::from_secs(2), ack_callback).unwrap();
     ///
     /// sleep(Duration::from_secs(2));
     /// ```
     #[inline]
-    pub fn emit_with_ack<F>(
+    pub fn emit_with_ack<F, E, D>(
         &mut self,
-        event: &str,
-        data: &str,
+        event: E,
+        data: D,
         timeout: Duration,
         callback: F,
     ) -> Result<()>
     where
-        F: FnMut(String) + 'static + Send + Sync,
+        F: FnMut(Payload) + 'static + Send + Sync,
+        E: Into<Event>,
```

Comment:
Uncle Bob says in Clean Code that no method needs more than 3 parameters. I don't know, but maybe some intermediate struct would make sense here? But then again, its just Uncle Bob Martin

Reply:
Mh, don't think that's an issue, when I look at some other socketio client APIs I don't see a different layout. As a user it would make more trouble to build up some intermediate struct that can't be used elsewhere.